pylint/checkers/variables.py

```diff
@@ -2198,7 +2198,9 @@ def _loopvar_name(self, node: astroid.Name) -> None:
         # the usage is safe because the function will not be defined either if
         # the variable is not defined.
         scope = node.scope()
-        if isinstance(scope, nodes.FunctionDef) and any(
+        # FunctionDef subclasses Lambda due to a curious ontology. Check both.
+        # See https://github.com/PyCQA/astroid/issues/291 (fix abandoned in a stale branch)
```

Comment:
Do you think we could merge this commit on main astroid ?

Reply:
I sort of like the idea, but it would be a very big breaking change. Everywhere in `pylint` (and I guess `astroid` also) that does any `isinstance` for lambda or function will need to be audited and have additional tests written. At the very least we could reopen the issue and add it to a milestone.

Reply:
Yeah it sounds like something that could go in astroid 3.0.

Reply:
For instance, my changes here in this PR would break!

Reply:
Thinking more, maybe the proposal to switch the parent/child classes makes no more sense than the current state of things. I don't have a good intuition for which one should be the parent. Maybe they should both inherit from `BaseFunction` (like `BaseContainer`) and break less code (just the code that checks for lambda and expects to also catch FunctionDef? but that seems rare ...)

Reply:
_Wow, oops_, that commit you linked looks like it does basically what I just said. So yeah, that's starting to look pretty good.

Reply:
```suggestion
        # See https://github.com/PyCQA/astroid/issues/291
        # TODO:  Revisit when astroid 3.0 include the change
```